### PR TITLE
Transition to J2735 ASN 2020, Dockerfile Optimization & Updated SCMS Files

### PR DESCRIPTION
**Pull request by dmccoystephenson**
# Transition to 2020
The ACM has transitioned to using the 2020 version of the J2735 ASN specification for compiling header and implementation files essential for message processing.

## Updated ASN.1 Compiler Repository
We have adopted Leidos' fork of the ASN.1 Compiler, accessible at https://www.github.com/mouse07410/asn1c. This change was necessary due to the 'WITH SUCCESSORS' keyword in the 2020 J2735 ASN specification, which is not supported by the original repository (https://www.github.com/vlm/asn1c) currently.

Adjustments to the dockerfiles were made to accommodate this switch.

## Modifications to doIt.sh Script
The doIt.sh script now directs to the 'asn1c_combined/j2735-asn-files/2020' directory, utilizing all *.asn files in that directory as input for the ASN.1 Compiler.

## SEMI ASN File Updates
The SEMI file has been updated to be compatible with J2735 2020. Structure references had to be changed and import statements were added.

This file has been renamed to differentiate it from the previous version of the file.

The changelog included at the top of the file has been updated, including the following:
- When the file was last updated
- Who updated the file last and what company they work for (this followed the convention already present)
- What got updated

## Unit Tests
Three new unit tests have been introduced for encoding TIMs with OdeTimPayload/OdeAsdPayload payload types and decoding a BSM. Each unit test is now encapsulated within its own TEST_CASE instead of being grouped under a single TEST_CASE.

Unit tests associated with encoding BSMs have been commented out, as they are not a desired feature of the ACM and were failing after the transition to mouse07410/asn1c.

### Test Data
There are some differences in the `TravelerDataFrame` structure between 2016 & 2020 that required some modifications to some test messages.

These include the following field renamings:
- sspTimRights -> notUsed
- duratonTime -> durationTime
- sspLocationRights -> notUsed1
- sspMsgRights1 -> notUsed2
- sspMsgRights2 -> notUsed3

## Testing
BSM Decoding & TIM Encoding have been verified to be working through manual testing. This involved spinning up the ACM & sending data to the relevant kafka topics.

## ODE Updates
Transitioning to J2735 2020 requires some modifications to the ODE as well. These have been included in the following PR:
https://github.com/usdot-jpo-ode/jpo-ode/pull/522

# Dockerfile Optimization
## Changes
Some changes have been made to the Dockerfiles with the aim to enhance performance and efficiency:

1. Base Image: The base image in the Dockerfiles has been switched from Ubuntu to Alpine, resulting in smaller and faster deployments.
1. Splitting Dockerfiles: The Dockerfiles are now divided into two images - the builder image and the runtime image. The project is built within the builder container, and only the necessary files are copied to a new container for deployment.

## Testing
The changes have been thoroughly tested with the following results:

1. Successful Compilation: The project compiles and starts up successfully.
1. Passing Unit Tests: All unit tests pass within the Alpine container.
1. Dev Cluster Deployment: When deployed to the development cluster, the data gets processed as expected.

These updates are expected to contribute to a more efficient and streamlined development and deployment process.

# Updated SCMS Files
## Changes
The purpose of the changes in this PR is to transition the ACM away from utilizing the 'scms-asn1' git submodule, opting instead to directly include the updated SCMS files within the project.

### Git Submodule Changes
The 'scms-asn1' git submodule has been removed.

### Updated SCMS Files
The SCMS files that were being used before were from 2018. These files are from 2021/2022 and include the following:

- EtsiTs103097ExtensionModule v1.0
- Ieee1609Dot2 v2.6
- Ieee1609Dot2BaseTypes v2.4
- Ieee1609Dot2Crl v3.2
- Ieee1609Dot2CrlBaseTypes v3.2

The 'Ieee1609Dot2.asn' file depends on a structure defined in the 'EtsiTs103097ExtensionModule.asn' file.

### Testing
- The unit tests have been verified to pass with these changes.
- The capability of the ACM to process a signed TIM for double-encoding has been verified.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,50 +1,57 @@
...
-
 VOLUME ["/asn1_codec_share"]
 
-# Add build tools.
-RUN apt-get update && apt-get install -y software-properties-common wget git make gcc-7 g++-7 gcc-7-base && update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-7 100 && update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-7 100
-
-# Install cmake.
-RUN apt-get install -y cmake
+# update the package manager
+RUN apk update
```
Comment (dan-du-car) [inline Dockerfile]: Not needed if using --no-cache below?

File: Dockerfile [hunk 2]
```diff
@@ -1,50 +1,57 @@
...
+    bison
 
 # Install pugixml
 ADD ./pugixml /asn1_codec/pugixml
 RUN cd /asn1_codec/pugixml && mkdir -p build && cd build && cmake .. && make && make install
 
 # Build and install asn1c submodule
 ADD ./asn1c /asn1_codec/asn1c
-RUN cd /asn1_codec/asn1c && aclocal && test -f configure || autoreconf -iv && ./configure && make && make install
+RUN cd asn1c && test -f configure || autoreconf -iv && ./configure && make && make install
```
Comment (dan-du-car) [inline Dockerfile:27]: Can we make this consistent with the pugxml install by installing it inside image directory /asn1_codec/asn1c?

File: Dockerfile [hunk 3]
```diff
@@ -1,50 +1,57 @@
...
 
-# Generate ASN.1 API.
+# Make generated files available to the build & compile example
 RUN export LD_LIBRARY_PATH=/usr/local/lib
 ADD ./asn1c_combined /asn1_codec/asn1c_combined
-ADD ./scms-asn1 /asn1_codec/scms-asn1
```
Comment (dan-du-car) [inline Dockerfile:37]: Why is this removed?

File: asn1c_combined/doIt.sh [hunk 4]
```diff
@@ -1,15 +1,28 @@
 #!/bin/bash
 
+# This script assumes that the J2735 files have already been generated.
+# To generate the J2735 files, run generate-files.sh with asn1c installed.
+
 export LD_LIBRARY_PATH=/usr/local/lib
 export CC=gcc
 
-asn1c -fcompound-names -gen-PER -gen-OER -pdu=all \
```
Comment (dan-du-car) [inline asn1c_combined/doIt.sh:6]: Is there a reason that you prefer to manually copy the generated source cpp and header files to automating this step during the docker build?

File: data/InputData.encoding.tim.odetimpayload.xml [hunk 5]
```diff
@@ -1 +1 @@
...
...><durationTime>1</durationTime><priority>0</priority><notUsed1>0</notUsed1><regions><GeographicalPath><name>Testing TIM</name><id><region>0</region><id>33</id></id><anchor><lat>412500807</lat><long>-1110093847</long><elevation>20206</elevation></anchor><laneWidth>700</laneWidth><directionality><both/></directionality><closedPath><false/></closedPath><direction>0000000000010100</direction><description><path><scale>0</scale><offset><ll><nodes><NodeLL><delta><node-LL3><lon>14506</lon><lat>31024</lat></node-LL3></delta></NodeLL><NodeLL><delta><node-LL3><lon>14568</lon><lat>30974</lat></node-LL3></delta></NodeLL><NodeLL><delta><node-LL3><lon>14559</lon><lat>30983</lat></node-LL3></delta></NodeLL><NodeLL><delta><node-LL3><lon>14563</lon><lat>30980</lat></node-LL3></delta></NodeLL><NodeLL><delta><node-LL3><lon>14562</lon><lat>30982</lat></node-LL3></delta></NodeLL></nodes></ll></offset></path></description></GeographicalPath></regions><notUsed2>0</notUsed2><notUsed3>0</notUsed3><content><advisory><SEQUENCE><item><itis>513</itis></item></SEQUENCE></advisory></content><url>null</url></TravelerDataFrame></dataFrames></TravelerInformation></value></MessageFrame></data></payload></OdeAsn1Data>
```
Comment (dan-du-car) [inline data/InputData.encoding.tim.odetimpayload.xml:1]: It would be helpful if we could update the README.md where each .xml in this data 
 folder generated/output to?

File: data/asd/asd.v3.mf.xml [hunk 6]
```diff
@@ -45,8 +45,8 @@
                         <direction>1010101010101010</direction>
                     </GeographicalPath>
                 </regions>
-                <sspMsgRights1>2</sspMsgRights1>
-                <sspMsgRights2>3</sspMsgRights2>
```
Comment (dan-du-car) [inline data/asd/asd.v3.mf.xml:49]: It would be helpful if you could update the README.md in this data folder to note how/where all files in asd folder generated/used?

File: data/examples/asn1c_input_tim.bin2xml [hunk 7]
```diff
@@ -97,8 +97,8 @@
                     </description>
                 </GeographicalPath>
             </regions>
-            <sspMsgRights1>2</sspMsgRights1>
```
Comment (dan-du-car) [inline data/examples/asn1c_input_tim.bin2xml:100]: Update the README.md how/where these xml examples generated or how they are used?

File: src/acm.cpp [hunk 8]
```diff
@@ -488,6 +488,10 @@ bool ASN1_Codec::configure() {
 
     // confluent cloud integration
     std::string kafkaType = getEnvironmentVariable("KAFKA_TYPE");
+    if (kafkaType == "") {
```
Comment (dan-du-car) [inline src/acm.cpp]: Are there any other kafka types other than local and confluent? If not, how about using an else statement after  `if (kafkaType == "CONFLUENT")` saying if confluent is not yet, it will by default use local kafka broker?

File: src/acm.cpp [hunk 9]
```diff
@@ -699,6 +703,11 @@ bool ASN1_Codec::make_loggers( bool remove_files ) {
     return true;
 }
 
+bool ASN1_Codec::make_loggers_testing() {
+    logger = std::make_shared<AcmLogger>("testlog");
```
Comment (dan-du-car) [inline src/acm.cpp]: Can we use make_unique instead? What is this hardcoded testlog string? Can we make the string a constant with a more descriptive name? Can we rename this function name to be more specific? It sounds like it makes loggers for testing? Maybe rename it to set_logger_name or something that tell what the function is about?

File: Dockerfile (hunk 2 above)
Comment (dmccoystephenson) [reply Dockerfile:27]: This is already getting installed into the /asn1_codec/asn1c directory.

The `WORKDIR /asn1_codec` instruction was executed prior to this. We are using a local reference to cd into the /asn1c_codec/asn1c directory on line 30. Should we change this to an absolute reference?

File: Dockerfile (hunk 3 above)
Comment (dmccoystephenson) [reply Dockerfile:37]: We are now including the SCMS files directly in the repository (in a subdirectory of the `asn1c_combined` directory), rather than separately in the scms-asn1 repository. The scms-asn1 git submodule was removed.

File: asn1c_combined/doIt.sh (hunk 4 above)
Comment (dmccoystephenson) [reply asn1c_combined/doIt.sh:6]: According to the header comments in the ASN J2735 2020 files, we are not permitted to redistribute the ASN J2735 2020 files. Due to this, we removed the ASN J2735 2020 files from source control and chose to have the developer retrieve the files themselves.

Once we removed the ASN files, the CI checks were failing because the automated step was attempting to use files that weren't there.

Moving to manually copying the generated source cpp & header files allowed us to keep the ASN files out of source control and it allowed the CI check to pass.

An additional benefit is that the build time has been reduced.

File: Dockerfile (hunk 1 above)
Comment (dmccoystephenson) [reply Dockerfile]: This has been removed.

File: data/InputData.encoding.tim.odetimpayload.xml (hunk 5 above)
Comment (dmccoystephenson) [reply data/InputData.encoding.tim.odetimpayload.xml:1]: Documentation has been added for which data files are used and where.

File: data/asd/asd.v3.mf.xml (hunk 6 above)
Comment (dmccoystephenson) [reply data/asd/asd.v3.mf.xml:49]: Documentation has been added for which data files are used and where. These don't appear to be referenced anywhere.

File: src/acm.cpp (hunk 8 above)
Comment (dmccoystephenson) [reply src/acm.cpp]: This has been refactored to improve clarity.

File: src/acm.cpp (hunk 9 above)
Comment (dmccoystephenson) [reply src/acm.cpp]: Apparently 'make_unique' is only available from C++14 onwards.

The logger setup code has been refactored to improve clarity.

**Comment by dmccoystephenson** [comment]
This PR is ready for re-review.

File: data/examples/asn1c_input_tim.bin2xml (hunk 7 above)
Comment (dmccoystephenson) [reply data/examples/asn1c_input_tim.bin2xml:100]: Documentation has been added for which data files are used and where. These don't appear to be referenced anywhere.

**Comment by dan-du-car** [comment]
Looks good. Thank you!
